### PR TITLE
perf(moe): optimize EPMoE unpermute

### PR DESCRIPTION
**Pull request by Junyi-99**
## Motivation

> **Stacked on #1577.**

The einsum put `top_k` on the VPU's sublane axis, which for top_k < 8 cost a full relayout copy of the gathered values.

```python
reshaped_intermediate = jnp.reshape(unsort_intermediate, (weights.shape[0], top_k, -1))
output = jnp.einsum("BKE,BK -> BE", reshaped_intermediate.astype(jnp.float32), weights_fp32)
```

Reshaping the index array rather than the values, and summing over `top_k` with an unrolled loop instead of an einsum, means no `(tokens,top_k, hidden)` tensor is built. 

Related PR: https://github.com/sgl-project/sglang-jax/pull/1058

## Modifications

`python/sgl_jax/srt/layers/moe.py` — in `EPMoE._unpermute`, reshape
`argsort_indices` (`int32[tokens*top_k]`, ~128 KB) into `(tokens, top_k)` and gather
column `k` directly, then accumulate over `range(top_k)` in float32.

`test_unpermute_sums_over_top_k`, parametrised over `top_k ∈ {1, 4, 8}`. (The existing test covers `top_k = 2` only)

## Accuracy Tests

For `top_k ≤ 2`: bit-identical.

For `top_k ≥ 4`: the unrolled sequential float32 accumulation differs from XLA's reduction order by about one bfloat16 ulp. Comparing `_unpermute` output against the parent commit on identical inputs, TPU v6e-1:

| shape | max abs delta | bit-identical |
|---|---:|---|
| Grok-2 `top_k=2 hidden=8192` T=2048 | 0 | yes |
| Grok-2 `top_k=2 hidden=8192` T=8192 | 0 | yes |
| Qwen3-235B `top_k=8 hidden=4096` T=2048 | 0.0078125 | no |
| Qwen3-235B `top_k=8 hidden=4096` T=8192 | 0.03125 | no |

Please flag if bit-identity at `top_k ≥ 4` is required.

```bash
python3 -m pytest python/sgl_jax/test/layers/test_moe_unpermute.py \
                  python/sgl_jax/test/kernels/moe_block_quant_test.py \
                  python/sgl_jax/test/test_moe_weight_loader_sharding.py -q
# 12 passed
```

## Benchmarking and Profiling

`EPMoE._unpermute` alone, TPU v6e-1, bf16, median of 20 runs.

| model shape                      | tokens |    parent |       this PR |   speedup |
| -------------------------------- | -----: | --------: | ------------: | --------: |
| Grok-2 `top_k=2 hidden=8192`     |   2048 |  510.8 us |  **187.3 us** | **2.73x** |
| Grok-2 `top_k=2 hidden=8192`     |   8192 | 2759.3 us | **1493.7 us** | **1.85x** |
| Qwen3-235B `top_k=8 hidden=4096` |   2048 |  860.8 us |  **756.3 us** |     1.14x |
| Qwen3-235B `top_k=8 hidden=4096` |   8192 | 3519.5 us | **3299.0 us** |     1.07x |

At top_k=2 the peak temp allocation halves (512.1 → 256.4 MiB at T=8192)

 At `top_k=8` the
`(tokens, 8, hidden)` reshape is already a free bitcast. Among currently supported models only Grok-2 has `top_k = 2`; the rest are 6–16 and land in the smaller-gain column.

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update. — n/a, internal perf change

https://claude.ai/code/session_01FStM6jUdJJWzzj4xCiXna6

### CODE REVIEW

**Review by Bob-Chen222 — DISMISSED** [review]

**Review by Bob-Chen222 — APPROVED** [review]

**Comment by Junyi-99** [comment]
Hi @cjx0709 @JamesBrianD, this one is approved and all checks are green, could one of you merge it when you have a moment? Thanks!
